modules/Channel.md

```diff
@@ -0,0 +1,112 @@
+	fmt.Println(msg)
+}
+```
+
+The module will have a generic `Channel` class (or record) to store the channel
+buffer and support operations on them.
+
+* The main functions that the module should have
+
+	* `init(type eltType, size : int)` : Specify the buffer data type and size
+	for creating a Channel.
+
+	* `send(value : eltType)` : This function waits for a space in the buffer
+	and	sends the `value` to the channel.
+
+	* `recv(out value : eltType) : bool` : This function waits until there is
+	something in the buffer and stores the received entity in `value`. It also
+	returns the status i.e., the operation is executed successfully or not.
+
+	* `close()` : This function is used to close the channel such that no more
+	values will be sent to it.
+
+* `iter these()` : An iterator to receive the values sent to the channel until
+it is closed.
+
+* `select` statements : This statement lets you wait for multiple channel
+operations and executes if anyone of them is ready for communication. `select`
+statements in Go are specifically designed to work with channels.
+Some key points are:
+
+	* `select` statements are commonly used in Go code when working with channels.
+	* The operations are atomic, i.e., checking a channel and performing `recv`
+		or `send` on it is considered as one operation.
+	* The cases are arranged in a random order to prevent starvation.
+
+For e.g.
+```
+// Select syntax in Go to wait on any one of the two channels for receiving
+// If every channel is blocked, the default case is executed.
+select {
+	case msg1 := <-chan1 :
+		fmt.Println("Received", msg1)
+	case msg2 := <-chan2 :
+		fmt.Println("Received", msg2)
+	default :
+		fmt.Println("None")
+}
+```
+
+```
+// Select statements can be used to implement try-send or try-recv
+// default case is executed when none of the cases are ready yet.
+select {
+	case val := <-chan1 :
+		fmt.Println("Received", val);
+	default :
+		fmt.Println("Channel not ready");
+}
+```
+
+### Some important questions related to the module-
+
+* Should this module be standard or package ?
+
+* What should be the name of the module ? Some names in other languages
+	* `chan` in Golang
+	* `crossbeam` in Rust
+
+* What should be the name of the channel data type? Should it be record or
+class ?
+	* Channel should always be passed as a reference therfore we could make
+	it a class or a record-wrapped class.
+
+* What should be the syntax of Channel `select` statements in Chapel ?
+Should we use Chapel's `select` statement for that? Some semantic differences
+between `select` in Chapel and Channels.
+
+	* Chapel's select is a conditional statement which is used to evaluate
+	cases against one statement specified after `select` keyword whereas
+	`select` in Channels work with multiple Channel operations.
+
+	* `select` for Channels can be blocking or non-blocking while `select`
+	in Chapel is always non-blocking.
+
+	* A Go-style `select` statement has no argument but a Chapel one does have an
```

Comment:
They certainly can be distinguished this way, but this `select` is more aligned with the `select` system call, whereas the other one we have right now is a conditional statement. So the idea behind using `select` as a conditional statement, I think, is the one that needs further deliberation because the first thought when someone familiar with system calls hears `select` he probably thinks about the system call one and not the conditional. Maybe deprecate the previous syntax to avoid future confusion?

Reply:
@mppf and I discussed this a bit a few weeks ago, and actually liked the juxtaposition of using the same keyword for both contexts, as they seemed similar "do this when this happens" and not likely to conflict directly in practice (one requires an expression that will be evaluated and matched against cases; the other has nothing to evaluate, and simply matches).  So I wouldn't shy away from it outright any more than we shied away from using `if` for both compiler-folded and normal conditionals; or mixing static type/param information with execution-time value information in argument lists.  But I agree that it's definitely a bit subtle, and that we should keep an eye on it to see whether it leads to confusion or has a greater potential for confusion than Michael or I are giving it credit for.

Reply:
Continuing with the above comment now that I've read the whole document:  For me a big question is "in the no-expression `select` statement [this new systems-y one we're talking about], what is the quality/property of the expressions that follow each of the `case` statements that makes them legal / reasonable?"  For example, in a normal `select` statement, I might say that the property is "each of the case statements has an expression whose type matches the type of the expression that followed the `select` keyword."  My intuition in this new case (as someone who hasn't done much of this event-driven systems programming before) is that perhaps each one has some sort of "system event" property (or type?  or ...?)?  Or is it even simpler than that—does each simply evaluate to true or false (or something similarly simple and primitive) and the evaluator just chooses a random case that evaluates to `true` to execute it?

The reason I ask this is that I'm curious whether there are other uses for this systems-y select statement that would cause us to want to use it apart from for these Go-like channels.  And if so, what a type or capability would have to do in order to be a legal condition.

One other thought on this general topic is that I find myself wondering about the role of asynchrony in this select statement given that Chapel also has a very strong story for asynchronous execution.  E.g., should the asynchrony be explicit in the statement somehow, either in or around the `select` keyword, or perhaps in the `when` clauses themselves?  Or is the synchronous vs. asynchronous behavior important enough that we should call it out with some sort of new keyword like `coselect` or `select begin` or ...?

Reply:
> what is the quality/property of the expressions that follow each of the `case` statements that makes them legal / reasonable?

I think the select statement could be extended to handle events (apart from channels), like a file read/write, operations on atomic/sync/single variables, etc. 




> does each simply evaluate to true or false (or something similarly simple and primitive) and the evaluator just chooses a random case that evaluates to `true` to execute it?

Suppose we are thinking like each case to be evaluated to `true` or `false`. It would be better described as the case which first evaluates to true gets executed. If there are more than one `true` cases, then yes, one of them is chosen randomly and gets executed.

Reply:
I would think that `coselect` / `select begin` would be creating tasks though? And they don't (they just block). (In the straightforward migration of the Go design, when there is an `otherwise`, it won't block, but if there isn't one, it will. But neither of these create new tasks).

Additionally it was my understanding that the atomicity guarantee of the Go-style `select` would present challenges if the interface is just to evaluate each case to `true` or `false`. I wonder if it would be better to think about each of these `select` statements as something more like a context manager (see #17965) . What I see in common there is that the return type of the expression (i.e. the type of `myResource` in `manage myResource in createSomeContex()`) isn't necessarily the type of the thing managing the context.

That leads me to thinking that the type of `a <~ b` (when used outside of a select) isn't necessarily the same as the type of the thing that the implementation of `select` is working with.

Reply:
@mppf / @Yudhishthira1406 : Sorry for the delayed response due to vacation.  To make sure I'm understanding:

* upon reaching a `select` statement, only one case at most will be executed, synchronously
  * if no cases apply and an `otherwise` exists, the `otherwise` will run; otherwise, the statement will block until a case does apply
  * if multiple cases apply, one is chosen randomly
* after that case is executed, execution will continue with the next statement after the `select` statement
* if the programmer wants multiple cases to be handled, they'd need to put the `select` within a loop
* if the programmer wanted asynchrony, they'd need to express that explicitly (e.g., prefixing the `select` with a `begin` or possibly within the resulting action, if that's legal?)

Have I got all that right?

Reply:
Yep, that's right.
I have made a [document](https://github.com/Yudhishthira1406/Chapel-Channel/blob/select-deq/SELECT.md) demonstrating how the compiler would understand the select statement (if we used a variation of Chapel's select statement). I think it was worth posting here.

Reply:
>  "in the no-expression select statement [this new systems-y one we're talking about], what is the quality/property of the expressions that follow each of the case statements that makes them legal / reasonable?"

I think this quality/property is
 1. The function called is some sort of blocking operation waiting for an external event
 2. The implementation of that function and/or of `select` supports that function.
 
At a high level, it seems fairly clear to me that `accept` / `send` and `recv` in Socket programming would make sense as select statements

``` chpl
select {
  when myTcpSock.accept() {
  }
  when myUdpSock.send(x1) {
  }
  when myUdpSock.recv(x2) {
  }
}
```

Here I would imagine that we use some sort of strategy like the Context Manager to get the `select` implementation able to work with various types/functions defined in the library.

Reply:
I suppose that one way to handle the plugability would be for these functions to `throw` if the operation is not handled. The `select` implementation will try them in a random order and execute the body of the first one that succeeds. I think the compiler could generate some kind of class / function object / etc to handle running these expressions.